src/Main.java

```diff
@@ -1,54 +1,32 @@
 import java.util.Random;
-import java.util.Scanner;
 
 public class Main {
 
-	static final int CODE_LENGTH = 4;
-	static final int MAX_SECRET = 10000;
-
 	public static void main(String[] args) {
-		// Set up code validation
-		Validator codeValidator = new Validator(CODE_LENGTH, false);
-
-		// Create secret
-		Random randomGenerator = new Random();
-		int secretKey = 0;
-		boolean validatedSecret = false;
-		while (!validatedSecret) {
-			secretKey = randomGenerator.nextInt(MAX_SECRET);
-			validatedSecret = codeValidator.isValidCode(String.valueOf(secretKey));
-		}
-		System.out.println("The secret is " + secretKey);
-		Digits secret = (Digits) CodeFactory.getInstance(secretKey);
-
-		Scanner input = new Scanner(System.in);
-
-		boolean gameIsWon = false;
-
-		while (!gameIsWon) {
-			// Get guesses from the user
-			System.out.println("Guess a 4-digit number: ");
-
-			String userInput = input.next();
-
-			if (codeValidator.isValidCode(userInput)) {
-				Digits guess = (Digits) CodeFactory.getInstance(Integer.parseInt(userInput));
-
-				if (guess.equals(secret)) {
-					gameIsWon = true;
-					System.out.println("Game should be finished");
-				} else {
-					System.out.println("Guess does not match the secret. Please try again.");
-				}
-			} else {
-				System.out.println("Invalid guess. Please try again.");
-			}
-			;
-
+		Reader reader = new Reader();
+		Printer printer = new Printer();
+
+		Secret secret = generateGameSecret();
+		printer.printInstruction("The secret is " + secret.getDigits());
```

Comment:
Who should know how a Secret is _rendered_ ? 
I suppose you have 3 options here.

1. The Secret
2. The Printer
3. Main

If you decided that a Secret is responsible for knowing how it renders itself in this text-based application, you might chose to override the `toString()` method available to all objects. 

If you decided that the printer should decide how it renders something, it should only require raw data as an input. Here, the Secret itself.

You have sort of decided "both" by giving the printer a ready formatted string, using select certain values from the Secret object. In this sense, I could argue that it's actually Main that now knows how to render a Secret.

If I no longer wish to print to stdout, but instead send the output as JSON to an API - you would have to make changes to Main. Main lives inside our central jigsaw piece. I would have expected changes for such a hypothetical feature to make changes to the "adapter" jigsaw pieces that we plug in either side of our diagram.